dlt/common/storages/transactions.py

```diff
@@ -0,0 +1,210 @@
+            # the mtime of the file, followed by the name, which is also time sortable
+            output[name] = (mtime.timestamp(), name)
+        return output
+
+    def read(self) -> t.Optional[bytes]:
+        """Reads data from the file."""
+        if self._fs.exists(self.path):
+            return self._fs.cat(self.path)
+        return None
+
+    def write(self, content: bytes) -> None:
+        """Writes data within the transaction."""
+        if not self._is_locked:
+            raise RuntimeError("Cannot write to a file without a lock.")
+        self._fs.pipe(self.path, content)
+
+    def rollback(self) -> None:
+        """Rolls back the transaction."""
+        if not self._is_locked:
+            raise RuntimeError("Cannot rollback a file without a lock.")
+        if self._original_contents is not None:
+            self._fs.pipe(self.path, self._original_contents)
+
+    def acquire_lock(self, blocking: bool = True, timeout: float = -1) -> bool:
+        """Acquires a lock on a path.
+
+        Acquire a lock, blocking or non-blocking.
+
+        When invoked with the blocking argument set to True (the default), block until
+        the lock is unlocked, then set it to locked and return True.
+
+        When invoked with the blocking argument set to False, do not block. If a call
+        with blocking set to True would block, return False immediately; otherwise, set
+        the lock to locked and return True.
+
+        When invoked with the floating-point timeout argument set to a positive value,
+        block for at most the number of seconds specified by timeout and as long as the
+        lock cannot be acquired. A timeout argument of -1 specifies an unbounded wait.
+        If blocking is False, timeout is ignored. The stdlib would raise a ValueError.
+
+        The return value is True if the lock is acquired successfully, False if
+        not (for example if the timeout expired).
+        """
+        if self._is_locked:
+            return True
+
+        self._fs.touch(self.lock_path)
+        locks = self._sync_locks()
+        _, active_lock = min(locks.values())
+        start = time.time()
+
+        while active_lock != self.lock_path:
+            if not blocking or (timeout > 0 and time.time() - start > timeout):
+                self._fs.rm(self.lock_path)
+                return False
+
+            time.sleep(random.random() + POLLING_INTERVAL)
+            locks = self._sync_locks()
+            if self.lock_path not in locks:
+                self._fs.touch(self.lock_path)
+                locks = self._sync_locks()
+
+            _, active_lock = min(locks.values())  # type: ignore
+
+        self._original_contents = self.read()
+        self._is_locked = True
+        if self._heartbeat is not None:
+            self._heartbeat.cancel()
+        self._heartbeat = self._make_heartbeat_thread()
+        self._heartbeat.start()
+        return True
+
+    def release_lock(self) -> None:
```

Comment:
should we add `release_lock` to the class destructor? then the lock will be released when garbage collected and heartbeat will stop. I'm not sure about that. IMO the only safe way to use locks is with a context manager or within other managed objects. not releasing lock is a programmer's mistake and maybe we should keep the current behavior

Reply:
I think it would be fair for the destructor to make an attempt at signalling the heartbeat to stop refreshing the mtime of the file but perhaps not to explicitly drop the lock. If a lock goes out of scope and hits the `gc`, it would be a non-obvious release of the lock before its TTL has expired.